codec/grpc/grpc_codec.go

```diff
@@ -1,33 +1,65 @@
 package grpc
 
-import "fmt"
+import (
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/mem"
+
+	// Guarantee that the built-in proto is called registered before this one
+	// so that it can be replaced.
+	_ "google.golang.org/grpc/encoding/proto"
+)
 
 // Name is the name registered for the proto compressor.
 const Name = "proto"
 
-type Codec struct{}
-
 type vtprotoMessage interface {
-	MarshalVT() ([]byte, error)
+	MarshalToVT(data []byte) (int, error)
 	UnmarshalVT([]byte) error
+	SizeVT() int
+}
+
+type Codec struct {
+	fallback encoding.CodecV2
 }
 
-func (Codec) Marshal(v interface{}) ([]byte, error) {
-	vt, ok := v.(vtprotoMessage)
-	if !ok {
-		return nil, fmt.Errorf("failed to marshal, message is %T (missing vtprotobuf helpers)", v)
+func (Codec) Name() string { return Name }
+
+func (c *Codec) Marshal(v any) (data mem.BufferSlice, err error) {
+	if m, ok := v.(vtprotoMessage); ok {
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, 0, size)
+			if _, err := m.MarshalToVT(buf[:0]); err != nil {
+				return nil, err
+			}
+			data = append(data, mem.SliceBuffer(buf))
```

Comment:
I was testing this patch and it returned empty (len=0, cap=size) buffers that failed some of my app tests. In my case the the underlying generated code for `MarshalToVT()` looks like this:
```go
func (m *Something) MarshalToVT(dAtA []byte) (int, error) {
	size := m.SizeVT()
	return m.MarshalToSizedBufferVT(dAtA[:size])
}
```

So I switched the desired `vtprotoMessage` interface to require `MarshalToSizedBufferVT` instead of `MarshalToVT`. We already have the result of `m.SizeVT()` anyway, which is not a simple thing to calculate, so this both fixes the bug and results in a more optimal code:
```suggestion
			buf := make([]byte, size)
			if _, err := m.MarshalToSizedBufferVT(buf); err != nil {
				return nil, err
			}
			data = mem.BufferSlice{mem.SliceBuffer(buf)}
```

Finally, there is no reason to use `append()` for the result. It will probably be optimized away by the Go compiler considering data is always going to be an empty slice here and below, but it's unnecessarily confusing when you read the code.